vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -113,25 +101,35 @@ def _action_callback_function(self, **inputs: Dict[str, Any]) -> Dict[str, Any]:
         logger.debug(f"Action inputs: {inputs}")
 
         # Invoking the action's function
-        return_value = self.function(**inputs) or {}
+        return_value = self.function(**inputs)
 
         # Action callback outputs
         outputs = list(ctx.outputs_grouping.keys())
         outputs.remove("action_finished")
 
-        # Validate number of outputs
-        self._validate_output_number(
-            outputs=outputs,
-            return_value=return_value,
-        )
-
-        # If return_value is a single element, ensure return_value is a list
-        if not isinstance(return_value, (list, tuple, dict)):
-            return_value = [return_value]
-        if isinstance(return_value, dict):
-            return {"action_finished": None, **return_value}
+        if return_value is None and len(outputs) == 0:
+            # Action has no outputs and returns None.
+            return_dict = {}
+        elif hasattr(return_value, "_asdict") and hasattr(return_value, "_fields"):
+            # return_value is a namedtuple.
+            if set(return_value._fields) != set(outputs):
+                raise ValueError(
+                    f"Action's returned fields {set(return_value._fields)} does not match the action's defined "
+                    f"outputs {set(outputs) if outputs else {}}."
```

Comment:
I refactored it to:
f"outputs {set(outputs) or {}}."


It's implemented like that because `print(set([]))` returns `"set()"` instead of `"{}"`, and I guess the `"{}"` is expected one.

Reply:
Ah ok, thanks for the explanation!